pkg/kube/multicluster/secretcontroller.go

```diff
@@ -142,7 +151,24 @@ func NewController(kubeclientset kube.Client, namespace string, clusterID cluste
 	return controller
 }
 
-func (c *Controller) AddHandler(h ClusterHandler) {
+type ComponentBuilder interface {
+	registerHandler(h handler)
+}
+
+// BuildMultiClusterComponent constructs a new multicluster component. For each cluster, the constructor will be called.
+// If the cluster is removed, the T.Close() method will be called.
+// Constructors MUST not do blocking IO; they will block other operations.
+// During a cluster update, a new component is constructed before the old one is removed for seamless migration.
+func BuildMultiClusterComponent[T ComponentConstraint](c ComponentBuilder, constructor func(cluster *Cluster) T) *Component[T] {
+	comp := &Component[T]{
+		constructor: constructor,
+		clusters:    make(map[cluster.ID]T),
+	}
+	c.registerHandler(comp)
+	return comp
+}
+
+func (c *Controller) registerHandler(h handler) {
```

Comment:
lock handlers?

Reply:
This is intended and matches the existing behavior -- I will add a comment about why